### PR TITLE
[FEATURE] generate typestubs

### PR DESCRIPTION
**Pull request by jpfeuffer**
Generates .pyi files with typing stubs next to the pyx files.

beta stadium

### CODE REVIEW

File: autowrap/ConversionProvider.py [hunk 1]
```diff
@@ -276,8 +300,41 @@ def output_conversion(self, cpp_type, input_cpp_var, output_py_var):
         return "%s = <%s>%s" % (output_py_var, cpp_type, input_cpp_var)
 
 
+class UnsignedIntegerConverter(TypeConverterBase):
+
+    """
+    wraps long and int. "long" base_type is converted to "int" by the
```
Comment (timosachsenberg) [inline autowrap/ConversionProvider.py:306]: is this relevant to the long unique ids we use?

File: autowrap/ConversionProvider.py [hunk 2]
```diff
@@ -1060,6 +1164,14 @@ def matches(self, cpp_type):
     def matching_python_type(self, cpp_type):
         return "list"
 
+    def matching_python_type_full(self, cpp_type):
+        tt, = cpp_type.template_args
+        try: # TODO if the template type is a typedef we currently have a problem
+            inner_conv = self.converters.get(tt)
+            return "List[%s]" % inner_conv.matching_python_type_full(tt)
+        except:
+            return "Any"
```
Comment (timosachsenberg) [inline autowrap/ConversionProvider.py:1173]: does this handle the template case? the comment "have a problem" could be improved

File: autowrap/Utils.py [hunk 3]
```diff
@@ -79,6 +84,9 @@ def compile_and_import(name, source_files, include_dirs=None, **kws):
         print("\n")
     for source_file in source_files:
         shutil.copy(source_file, tempdir)
+        stub = source_file[:-4]+".pyi"
```
Comment (timosachsenberg) [inline autowrap/Utils.py:87]: gues the -4 is save?

File: tests/test_code_generator_libcpp.py [hunk 4]
```diff
@@ -176,9 +176,6 @@ def test_libcpp():
     assert list(out.items()) == [(1, 2.0)]
 
     out = t.process13(libcpp.EEE.A, 2)
-    print (list(out.items()))
-    print ([(libcpp.EEE.A, 2)])
-    print (list(out.items()) == [(libcpp.EEE.A, 2)])
     assert list(out.items()) == [(libcpp.EEE.A, 2)]
```
Comment (timosachsenberg) [inline tests/test_code_generator_libcpp.py:179]: can we test the type stubs / typelist generation somehow?

File: autowrap/CodeGenerator.py [hunk 5]
```diff
@@ -678,9 +717,13 @@ def _create_fun_decl_and_input_conversion(self, code, py_name, method, is_free_f
         """
         args = augment_arg_names(method)
 
+        if method.name == "process7":
```
Comment (timosachsenberg) [inline autowrap/CodeGenerator.py:720]: remove debug code?

**Comment by timosachsenberg** [comment]
Awesome. A few questions:
- Instead of having a pyi file could the type stubs also be generated inline?
I think it is considered the preferred way "if possible" TM.
- Does this have any impact on pylance? e.g., I would guess that it might now display the type stubs
but maybe not the documentation?

**Comment by jpfeuffer** [comment]
Hi,
1) No, I think Cython cannot handle that. https://github.com/cython/cython/issues/2587#issuecomment-881900062
2) Here are examples, F1 or hover for doc, Ctrl+P for params.
Bildschirmfoto 2022-01-18 um 11 25 10
Bildschirmfoto 2022-01-18 um 11 24 56
Bildschirmfoto 2022-01-18 um 11 24 48

File: autowrap/ConversionProvider.py (hunk 1 above)
Comment (jpfeuffer) [reply autowrap/ConversionProvider.py]: Don't know. But this was like this before. I just added a new converter that also checks positivity.

File: autowrap/ConversionProvider.py (hunk 2 above)
Comment (jpfeuffer) [reply autowrap/ConversionProvider.py]: seems to work

File: autowrap/Utils.py (hunk 3 above)
Comment (jpfeuffer) [reply autowrap/Utils.py:86]: made it safer

File: tests/test_code_generator_libcpp.py (hunk 4 above)
Comment (jpfeuffer) [reply tests/test_code_generator_libcpp.py:179]: have to think about it. probably. you could at least compare the resulting files
